src/common/utils/map_iter.h

```diff
@@ -19,17 +19,19 @@
 #ifndef SRC_COMMON_UTILS_MAP_ITER_H_
 #define SRC_COMMON_UTILS_MAP_ITER_H_
 
+#include <algorithm>
 #include <vector>
 
 namespace raksha::utils {
 
-template <typename T, typename U, typename F>
-std::vector<U> MapIter(const std::vector<T>& input, F f) {
+// Returns a vector that is generating by applying `map_fn: (C::value_type ->
+// U)` to every elemnent of the input container `C`.
+template <typename U, typename C, typename F>
+std::vector<U> MapIter(const C& input, F map_fn) {
```

Comment:
I think having `MapIter` accept a `const` reference parameter could be a bit of a mistake. It prevents copy elision from making the result of `MapIter` be constructed directly into the

Reply:
Note that this shift would also make it preferred to have lambdas accepting an element type `E` to have their parameter type be `E` rather than `const E &`.

Reply:
> I don't think it will. The guidance from the style guide is to, whenever we have a function that is performing a copy of a parameter in its body, take the parameter by value instead of by reference to move the copy to the callsite to open the door for copy elision.

We are not always making a copy of the input container `C`. Rather, we usually transform it to something else. Won't we make unnecessary copies in such cases? 

A couple of other concerns with taking the container `C` by value: 
 - If we really want to move the values, we should be able to use `std::transform` directly to do an in-place update.
 - Moreover, if we really want to take advantage of the `move_iterator`, the `map_fn` also needs to take its argument by value. 
 
IMO, all of this unnecessary complicates the API. Think of this as a `map` in functional programming, where you map one sequence of values to another.

Reply:
Just for context, I was really going for an incremental change here. Further, after discovering `absl` helpers, I have started thinking if we should even have this API.  That said, I understand the spirit of your concern, but I think it will make the common case more expensive. It is possible that I am misunderstanding something here. Let us sync in person. 

> The body of this map function uses std::transform to read each element from the input and write a corresponding element into the result local, which is then returned. That means that the body of this function makes a copy of its input.

Given that we are using iterators, we won't be making a copy of the input if we use `std::transform`.  Am I wrong? The possible implementations in https://en.cppreference.com/w/cpp/algorithm/transform suggests no copies are made as well.

Reply:
Gogul and I talked offline. I hadn't taken into account that we are often considering only a few fields from the input vector when performing a `map`, so my argument about the equivalence of copies doesn't hold up as well as I thought.